Support new prefetching methodology for stride prefetcher

SabaJamilan opened the pull request.

hlitz commented on src/cache_arrays.cpp:
```diff
@@ -71,7 +73,7 @@ int32_t SetAssocArray::lookup(const Address lineAddr, const MemReq* req, bool up
 			profPrefSavedCycles.inc(array[id].availCycle - array[id].startCycle);
 		    }
 		}
-                //line is in flight, compensate for potential OOO
```
NIT: no reason for this, please revert

hlitz commented on src/cache_arrays.cpp:
```diff
@@ -94,7 +96,6 @@ int32_t SetAssocArray::lookup(const Address lineAddr, const MemReq* req, bool up
                         profPrefHit.inc();
                     }
                     array[id].prefetch = false;
-
```
NIT: no reason for this, please revert

hlitz commented on src/cache_arrays.cpp:
```diff
@@ -115,7 +116,6 @@ uint32_t SetAssocArray::preinsert(const Address lineAddr, const MemReq* req, Add
     uint32_t candidate = rp->rankCands(req, SetAssocCands(first, first+assoc));
 
     *wbLineAddr = array[candidate].addr;
-
```
NIT: no reason for this, please revert

hlitz commented on src/cache_arrays.cpp:
```diff
@@ -26,10 +26,12 @@
 #include "cache_arrays.h"
 #include "hash.h"
 #include "repl_policies.h"
+#include <inttypes.h>
```
not clear why these are needed. There should be no changes to cache_arrays.cpp in this pull request.

hlitz commented on src/prefetcher.cpp:
```diff
@@ -69,19 +163,40 @@ void StreamPrefetcher::initStats(AggregateStat* parentStat) {
     profStrideSwitches.init("strideSwitches", "Predicted stride switches"); s->append(&profStrideSwitches);
     profLowConfAccs.init("lcAccs", "Low-confidence accesses with no prefetches"); s->append(&profLowConfAccs);
     parentStat->append(s);
+     prof_emitted_prefetches_.init("pff", "Total prefetches emitted");
```
NIT: identation

hlitz commented on src/prefetcher.h:
```diff
@@ -58,7 +65,7 @@ class SatCounter {
  * FIXME: For now, mostly hardcoded; 64-line entries (4KB w/64-byte lines), fixed granularities, etc.
  * TODO: Adapt to use weave models
  */
-class StreamPrefetcher : public BaseCache {
```
NIT: leave spacing

hlitz commented on src/prefetcher.cpp:
```diff
@@ -26,36 +26,130 @@
 //#define DBG(args...) info(args)
 #define DBG(args...)
 
+void StreamPrefetcher::postInit() {
```
why is this copied from basecache? This class should just inherit from basecache.

SabaJamilan commented:
Hi Heiner,
I removed the nits that you mentioned before. Also, the new stream prefetcher inherits postInit, setParent, ... from the cache prefetcher class.
I added the changes to trace_init.cpp and I think it would be work now on your side.
Thanks,
Saba
